rust/src/types/execution_context.rs

```diff
@@ -0,0 +1,121 @@
+use evmc_vm::{
+    AccessStatus, Address, ExecutionContext, ExecutionMessage, ExecutionResult, ExecutionTxContext,
+    StorageStatus, Uint256,
+};
+
+#[cfg_attr(test, mockall::automock)]
+pub trait ExecutionContextTrait {
+    /// Retrieve the transaction context.
+    fn get_tx_context(&self) -> &ExecutionTxContext;
+
+    /// Check if an account exists.
+    fn account_exists(&self, address: &Address) -> bool;
+
+    /// Read from a storage key.
+    fn get_storage(&self, address: &Address, key: &Uint256) -> Uint256;
+
+    // Set value of a storage key.
+    fn set_storage(&mut self, address: &Address, key: &Uint256, value: &Uint256) -> StorageStatus;
+
+    /// Get balance of an account.
+    fn get_balance(&self, address: &Address) -> Uint256;
+
+    /// Get code size of an account.
+    fn get_code_size(&self, address: &Address) -> usize;
+
+    /// Get code hash of an account.
+    fn get_code_hash(&self, address: &Address) -> Uint256;
+
+    /// Copy code of an account.
+    fn copy_code(&self, address: &Address, code_offset: usize, buffer: &mut [u8]) -> usize;
+
+    /// Self-destruct the current account.
+    fn selfdestruct(&mut self, address: &Address, beneficiary: &Address) -> bool;
+
+    /// Call to another account.
+    fn call(&mut self, message: &ExecutionMessage) -> ExecutionResult;
+
+    /// Get block hash of an account.
+    fn get_block_hash(&self, num: i64) -> Uint256;
+
+    /// Emit a log.
+    fn emit_log(&mut self, address: &Address, data: &[u8], topics: &[Uint256]);
+
+    /// Access an account.
+    fn access_account(&mut self, address: &Address) -> AccessStatus;
+
+    /// Access a storage key.
+    fn access_storage(&mut self, address: &Address, key: &Uint256) -> AccessStatus;
+
+    /// Read from a transient storage key.
+    fn get_transient_storage(&self, address: &Address, key: &Uint256) -> Uint256;
+
+    /// Set value of a transient storage key.
+    fn set_transient_storage(&mut self, address: &Address, key: &Uint256, value: &Uint256);
+}
+
+impl<'a> ExecutionContextTrait for ExecutionContext<'a> {
```

Comment:
This implementation seems a little superfluous.
Would it be possible to implement `ExecutionContextTrait` for `ExecutionContext` where the old methods are declared and avoid the forwarding?

Reply:
no because ExecutionContext comes from evmc